### ISSUE TITLE
Validate proptype definitions sooner

### ISSUE THREAD

**Issue by zpao** [facebook/react#5468]
Right now we validate some of our proptype definitions, but we don't surface the issue until that component class is consumed. We should instead validate these, in `__DEV__` only as warnings when the class in created. This means that a call to `React.PropTypes.onOf('foo')` results in an immediate warning logged.

This is a relatively contained bit of code so would be a good candidate for somebody looking to start contributing to React.

**Comment by antoaravinth** [comment]
@zpao : If I'm correct, this is already done over here right : 

https://github.com/facebook/react/blob/master/src/isomorphic/ReactIsomorphic.js#l30

We are already checking in `__DEV__` mode and if so pointing `createElement` as `ReactElementValidator.createElement` which actually does calls the `validatePropTypes` function. 

So I didn't quite get this issue. Probably I didn't get these words properly from your text : "but we don't surface the issue until that component class is consumed."

**Comment by zpao** [comment]
@antoaravinth the code you linked validates proptypes on element creation. So assuming a test case like

```js
var MyComponent = React.createClass({
  propTypes: {
    foo: React.PropTypes.oneOf(1, 2)
  }
  render() { return null; }
}
```

I want that to warn right away. I've written an invalid proptype (`oneOf` must take an array).

But the current behavior is not to do that. Instead you won't know that the component has defined proptypes improperly until you do `ReactDOM.render(<MyComponent />)` (this is what I mean by the class being consumed - poor phrasing on my part)

**Comment by antoaravinth** [comment]
@zpao Now I got what you meant. Anyways looks like rppc have given an PR. Thanks for your reply.

**Comment by freshman88** [comment]
is that `React.PropTypes.onOf('foo')` or `React.PropTypes.oneOf('foo')`

**Comment by iam-peekay** [comment]
@zpao I'm very interested in contributing to React. Is this an issue that still needs to be worked on?

**Comment by jimfb** [comment]
@iam-peekay Looks like this one is nearly done (https://github.com/facebook/react/pull/5476).  If you're looking for a good getting-started issue, https://github.com/facebook/react/issues/5773 should be relatively easy and I don't think anyone has started it yet.

**Comment by iam-peekay** [comment]
@jimfb thanks! I'll take a look

**Comment by troydemonbreun** [comment]
This is my first contribution. Trying to keep my PR up to date, what process is most friendly to the FB devs? (didn't realize rebasing would cause noise on this thread)

**Comment by troydemonbreun** [comment]
#6316 is the valid PR for this.  This is my first contribution.

Sorry for the terrible commit noise.  I made the initial mistake of doing a PR on my master, and being new to github, didn't realize this would happen.

I started over with a new master and now I'll branch per PR.

**Comment by troydemonbreun** [comment]
I think @rppc should get this one.  @rppc feel free to copy the code from #6316 into your own PR, it is essentially the same code.  I thought your PR has been rejected since it had been untouched for around 2 months.  I'm still learning how things work here and thought you would get notifications of all that happened on this thread.  Sorry for stepping on your work.

**Comment by rppc** [comment]
@troydemonbreun That's OK, really. I had thought about pinging @zpao about this after a while, but I've been quite busy and kept forgetting. Just let it go, no hard feelings!

**Comment by gaearon** [comment]
:+1: Glad you resolved this. It happens :smile: . No worries please.

**Comment by troydemonbreun** [comment]
@rppc Thank you!

**Comment by gaearon** [comment]
Fixed in #6316
